flink-runtime/src/main/java/org/apache/flink/runtime/deployment/TaskDeploymentDescriptorFactory.java

```diff
@@ -0,0 +1,262 @@
+			new TaskDeploymentDescriptor.Offloaded<>(taskInfo.right());
+	}
+
+	private List<InputGateDeploymentDescriptor> createInputGateDeploymentDescriptors(ResourceID location) {
+		List<InputGateDeploymentDescriptor> inputGates = new ArrayList<>(inputEdges.length);
+
+		for (ExecutionEdge[] edges : inputEdges) {
+			ShuffleDescriptor[] consumedPartitions = getConsumedPartitionShuffleDescriptors(edges);
+			// If the produced partition has multiple consumers registered, we
+			// need to request the one matching our sub task index.
+			// TODO Refactor after removing the consumers from the intermediate result partitions
+			int numConsumerEdges = edges[0].getSource().getConsumers().get(0).size();
+
+			int queueToRequest = subtaskIndex % numConsumerEdges;
+
+			IntermediateResult consumedIntermediateResult = edges[0].getSource().getIntermediateResult();
+			IntermediateDataSetID resultId = consumedIntermediateResult.getId();
+			ResultPartitionType partitionType = consumedIntermediateResult.getResultType();
+
+			inputGates.add(new InputGateDeploymentDescriptor(
+				resultId,
+				partitionType,
+				queueToRequest,
+				consumedPartitions,
+				location));
+		}
+
+		return inputGates;
+	}
+
+	private ShuffleDescriptor[] getConsumedPartitionShuffleDescriptors(ExecutionEdge[] edges) {
+		ShuffleDescriptor[] shuffleDescriptors = new ShuffleDescriptor[edges.length];
+		// Each edge is connected to a different result partition
+		for (int i = 0; i < edges.length; i++) {
+			shuffleDescriptors[i] =
+				getConsumedPartitionShuffleDescriptor(edges[i], allowLazyDeployment);
+		}
+		return shuffleDescriptors;
+	}
+
+	public static ShuffleDescriptor getConsumedPartitionShuffleDescriptor(
+			ExecutionEdge edge,
+			boolean allowUnknownPartitions) {
+		IntermediateResultPartition consumedPartition = edge.getSource();
+		Execution producer = consumedPartition.getProducer().getCurrentExecutionAttempt();
+
+		ExecutionState producerState = producer.getState();
+		Optional<ResultPartitionDeploymentDescriptor> consumedPartitionDescriptor =
+			producer.getProducedPartition(consumedPartition.getPartitionId());
+
+		ResultPartitionID consumedPartitionId = new ResultPartitionID(
+			consumedPartition.getPartitionId(),
+			producer.getAttemptId());
+
+		return getConsumedPartitionShuffleDescriptor(
+			consumedPartitionId,
+			consumedPartition.getResultType(),
+			consumedPartition.isConsumable(),
+			producerState,
+			allowUnknownPartitions,
+			consumedPartitionDescriptor.orElse(null));
+	}
+
+	@VisibleForTesting
+	static ShuffleDescriptor getConsumedPartitionShuffleDescriptor(
+			ResultPartitionID consumedPartitionId,
+			ResultPartitionType resultPartitionType,
+			boolean isConsumable,
+			ExecutionState producerState,
+			boolean allowUnknownPartitions,
+			@Nullable ResultPartitionDeploymentDescriptor consumedPartitionDescriptor) {
+		// The producing task needs to be RUNNING or already FINISHED
+		if ((resultPartitionType.isPipelined() || isConsumable) &&
```

Comment:
As far as I remember we take into account `hasDataProduced` in `IntermediateResultPartition.isConsumable` for `ExecutionVertex.checkInputDependencyConstraints` introduced in #7255. Here `hasDataProduced` is not relevant for known partition. This might be up to discussion.